tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala

```diff
@@ -0,0 +1,594 @@
+                mkTypeDecl(ExactRef(decl.body.ID), name, tt, terminal)
+              } catch {
+                case e: Type1ParseError =>
+                  throw new TypingInputException(s"Parser error in type annotation of $name: ${e.msg}")
+              }
+          }
+        }
+
+      case _ =>
+        throw new TypingInputException(s"""Error in ${decl.name}: Expected /\ AssumeType(varName, "typeString") /\ ...""")
+    }
+  }
+
+  // TODO: a long string of translation rules. Can we decompose it?
+  def apply(ex: TlaEx): EtcExpr = ex match {
+    case NameEx(name) =>
+      // x becomes x
+      mkName(ExactRef(ex.ID), name)
+
+    //*********************************************** LITERALS **************************************************
+    case ValEx(TlaInt(_)) =>
+      // an integer literal becomes IntT1
+      mkConst(ExactRef(ex.ID), IntT1())
+
+    case ValEx(TlaBool(_)) =>
+      // a Boolean literal becomes BoolT1
+      mkConst(ExactRef(ex.ID), BoolT1())
+
+    case ValEx(TlaStr(_)) =>
+      // a string literal becomes StrT1
+      mkConst(ExactRef(ex.ID), StrT1())
+
+    case ValEx(TlaReal()) =>
+      // a real literal becomes RealT1
+      mkConst(ExactRef(ex.ID), RealT1())
+
+    case ValEx(TlaIntSet) =>
+      // the set of all integers is SetT1(IntT1)
+      mkConst(ExactRef(ex.ID), SetT1(IntT1()))
+
+    case ValEx(TlaNatSet) =>
+      // the set of all naturals is SetT1(IntT1)
+      mkConst(ExactRef(ex.ID), SetT1(IntT1()))
+
+    case ValEx(TlaRealSet) =>
+      // the set of all reals is SetT1(RealT1)
+      mkConst(ExactRef(ex.ID), SetT1(RealT1()))
+
+    case ValEx(TlaBoolSet) =>
+      // the set of all Booleans is SetT1(BoolT1)
+      mkConst(ExactRef(ex.ID), SetT1(BoolT1()))
+
+    case ValEx(TlaStrSet) =>
+      // the set of all strings is SetT1(StrT1)
+      mkConst(ExactRef(ex.ID), SetT1(StrT1()))
+
+    //**************************************** EMPTY SETS AND SEQUENCES ***********************************************
+    case OperEx(TypingOper.emptySet, ValEx(TlaStr(elemTypeText))) =>
+      try {
+        val elemType = type1Parser(elemTypeText)
+        mkConst(ExactRef(ex.ID), SetT1(elemType))
+      } catch {
+        case e: Type1ParseError =>
+          throw new TypingInputException(s"Parser error in Typing!EmptySet($elemTypeText): ${e.msg}")
+      }
+
+    case OperEx(TypingOper.emptySeq, ValEx(TlaStr(elemTypeText))) =>
+      try {
+        val elemType = type1Parser(elemTypeText)
+        mkConst(ExactRef(ex.ID), SeqT1(elemType))
+      } catch {
+        case e: Type1ParseError =>
+          throw new TypingInputException(s"Parser error in Typing!EmptySeq($elemTypeText): ${e.msg}")
+      }
+
+    //******************************************** GENERAL OPERATORS **************************************************
+    case OperEx(op, args @ _*) if op == TlaOper.eq || op == TlaOper.ne =>
+      // x = y, x /= y
+      val opsig = OperT1(Seq(VarT1("a"), VarT1("a")), BoolT1())
```

Comment:
Why does this hard code in type variable name `"a"`? Couldn't this conflict with other type variables in scope? Also, why don't we get the actual types of the arguments here in stead of just positing variables? Wouldn't this fail to produce the desired type checking error if someone passed in, e.g., `1 = "1"`, since it doesn't actually check whether the operands are the same type?

Reply:
All signatures have simple variable names, for convenience and simplicity. They get renamed in the type checking code.

Reply:
Ok. This part seems more significant:

> Wouldn't this fail to produce the desired type checking error if someone passed in, e.g., 1 = "1", since it doesn't actually check whether the operands are the same type?